File: leetcode_75/Stack/python/394.py
from collections import deque


class Solution:
    def decodeString(self, s: str) -> str:
        digit_stack = deque()
        letter_stack = deque()
        ans = ""
        last_symbol = ""
        for c in s:
            if c.isdigit():
                if not last_symbol.isdigit():
                    digit_stack.append(c)
                else:
                    digit_stack[-1] += c

            elif c == "[":
                letter_stack.append("")

            elif c.isalpha():
                if len(letter_stack) > 0:
                    letter_stack[-1] += c
                else:
                    ans += c

            elif c == "]":
                digit = int(digit_stack.pop())
                letter = letter_stack.pop()
                if len(letter_stack) > 0:
                    letter_stack[-1] += letter * digit
                else:
                    ans += letter * digit

            last_symbol = c

        return ans


solution = Solution()
print(solution.decodeString("100[leetcode]"))

File: leetcode_75/Stack/cpp/394.cpp
#include <iostream>
#include <string>
#include <stack>
#include <cctype>

using namespace std;

class Solution
{
public:
    string multiplyString(const string &str, int count)
    {
        string result = "";
        for (int i = 0; i < count; ++i)
        {
            result += str;
        }
        return result;
    }

    string decodeString(string s)
    {
        stack<string> digit_stack;
        stack<string> letter_stack;
        string ans;
        char last_symbol;

        for (auto c : s)
        {
            if (isdigit(c))
            {
                if (isdigit(last_symbol))
                {
                    digit_stack.top() = digit_stack.top() + c;
                }
                else
                {
                    digit_stack.push(string(1, c));
                }
            }

            else if (c == '[')
            {
                letter_stack.push("");
            }

            else if (isalpha(c))
            {
                if (letter_stack.size() > 0)
                {
                    letter_stack.top() = letter_stack.top() + c;
                }
                else
                {
                    ans += c;
                }
            }

            else if (c == ']')
            {
                int digit = stoi(digit_stack.top());
                digit_stack.pop();
                string letter = letter_stack.top();
                letter_stack.pop();

                if (letter_stack.size() > 0)
                {
                    letter_stack.top() += this->multiplyString(letter, digit);
                }
                else
                {
                    ans += this->multiplyString(letter, digit);
                }
            }

            last_symbol = c;
        }

        return ans;
    }
};